# Lab | Data Structuring and Combining Data

For Challenge 1, you can either use data which has already been cleaned (option 1) or clean it yourself (option 2).

## Challenge 1: Combining Data (Option 1)

In this challenge, we will be working with the customer data from an insurance company, as we did in the two previous labs. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

But this time, we got new data, which can be found in the following 2 CSV files located at the following links:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2_clean.csv
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3_clean.csv

Combine the data from the three dataframes into a single dataframe, named "customer_data", using appropriate merging, concatenating, and joining techniques.

Verify that the customer_data dataframe contains all the rows and columns from the three original dataframes.

## Challenge 1: Combining & Cleaning Data (Option 2)

If in the previous lab you created your cleaning and formatting function, or if you want to do it now, instead of using the two clean files provided above, you can use your function to clean and format the data in these two raw files below:

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv

Observation:
- One option is to first combine the three datasets and then apply the cleaning function to the new combined dataset
- Another option would be to read the clean file you saved in the previous lab, and just clean the two new files and concatenate the three clean datasets

In [29]:
! ls

README.md
__pycache__
dw_labs_data
file1_new.csv
file2_cleaned.csv
file3_cleaned.csv
functions.py
lab-dw-data-structuring-and-combining.ipynb


In [30]:
# Your code goes here
import pandas as pd
import functions

df1 = pd.read_csv("file1_new.csv")
df2 = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2_clean.csv")
df3 = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3_clean.csv")


In [31]:
df2=functions.cleaning_column_names(df2)

display(df2)

customer       state gender             education  \
0      RB50392  Washington    NaN                Master   
1      QZ44356     Arizona      F              Bachelor   
2      AI49188      Nevada      F              Bachelor   
3      WW63253  California      M              Bachelor   
4      GA49547  Washington      M  High School or Below   
...        ...         ...    ...                   ...   
12069  LA72316  California      M              Bachelor   
12070  PK87824  California      F               College   
12071  TD14365  California      M              Bachelor   
12072  UP19263  California      M               College   
12073  Y167826  California      M               College   

      customer_lifetime_value   income  monthly_premium_auto  \
0                         NaN      0.0                1000.0   
1                  697953.59%      0.0                  94.0   
2                 1288743.17%  48767.0                 108.0   
3                  764586.18%      0.0                 106.0   
4                  536307.65%  36357.0                  68.0   
...                       ...      ...                   ...   
12069             23405.98798  71941.0                  73.0   
12070             3096.511217  21604.0                  79.0   
12071             8163.890428      0.0                  85.0   
12072             7524.442436  21941.0                  96.0   
12073             2611.836866      0.0                  77.0   

      number_of_open_complaints     policy_type  vehicle_class  \
0                        1/0/00   Personal Auto  Four-Door Car   
1                        1/0/00   Personal Auto  Four-Door Car   
2                        1/0/00   Personal Auto   Two-Door Car   
3                        1/0/00  Corporate Auto            SUV   
4                        1/0/00   Personal Auto  Four-Door Car   
...                         ...             ...            ...   
12069                         0   Personal Auto  Four-Door Car   
12070                         0  Corporate Auto  Four-Door Car   
12071                         3  Corporate Auto  Four-Door Car   
12072                         0   Personal Auto  Four-Door Car   
12073                         0  Corporate Auto   Two-Door Car   

       total_claim_amount  
0                2.704934  
1             1131.464935  
2              566.472247  
3              529.881344  
4               17.269323  
...                   ...  
12069          198.234764  
12070          379.200000  
12071          790.784983  
12072          691.200000  
12073          369.600000  

[12074 rows x 11 columns]

In [32]:
df3=functions.cleaning_column_names(df3)
display(df3)

customer       state gender             education  \
0      RB50392  Washington    NaN                Master   
1      QZ44356     Arizona      F              Bachelor   
2      AI49188      Nevada      F              Bachelor   
3      WW63253  California      M              Bachelor   
4      GA49547  Washington      M  High School or Below   
...        ...         ...    ...                   ...   
12069  LA72316  California      M              Bachelor   
12070  PK87824  California      F               College   
12071  TD14365  California      M              Bachelor   
12072  UP19263  California      M               College   
12073  Y167826  California      M               College   

      customer_lifetime_value   income  monthly_premium_auto  \
0                         NaN      0.0                1000.0   
1                  697953.59%      0.0                  94.0   
2                 1288743.17%  48767.0                 108.0   
3                  764586.18%      0.0                 106.0   
4                  536307.65%  36357.0                  68.0   
...                       ...      ...                   ...   
12069             23405.98798  71941.0                  73.0   
12070             3096.511217  21604.0                  79.0   
12071             8163.890428      0.0                  85.0   
12072             7524.442436  21941.0                  96.0   
12073             2611.836866      0.0                  77.0   

      number_of_open_complaints     policy_type  vehicle_class  \
0                        1/0/00   Personal Auto  Four-Door Car   
1                        1/0/00   Personal Auto  Four-Door Car   
2                        1/0/00   Personal Auto   Two-Door Car   
3                        1/0/00  Corporate Auto            SUV   
4                        1/0/00   Personal Auto  Four-Door Car   
...                         ...             ...            ...   
12069                         0   Personal Auto  Four-Door Car   
12070                         0  Corporate Auto  Four-Door Car   
12071                         3  Corporate Auto  Four-Door Car   
12072                         0   Personal Auto  Four-Door Car   
12073                         0  Corporate Auto   Two-Door Car   

       total_claim_amount  
0                2.704934  
1             1131.464935  
2              566.472247  
3              529.881344  
4               17.269323  
...                   ...  
12069          198.234764  
12070          379.200000  
12071          790.784983  
12072          691.200000  
12073          369.600000  

[12074 rows x 11 columns]

In [33]:
df_final = pd.concat([df1,df2,df3], axis=0).reset_index(drop=True)
df_final

customer       state gender             education  \
0      RB50392  Washington      U                Master   
1      QZ44356     Arizona      F              Bachelor   
2      AI49188      Nevada      F              Bachelor   
3      WW63253  California      M              Bachelor   
4      GA49547  Washington      M  High School or Below   
...        ...         ...    ...                   ...   
25214  LA72316  California      M              Bachelor   
25215  PK87824  California      F               College   
25216  TD14365  California      M              Bachelor   
25217  UP19263  California      M               College   
25218  Y167826  California      M               College   

      customer_lifetime_value   income  monthly_premium_auto  \
0                      588174      0.0                1000.0   
1                      697953      0.0                  94.0   
2                     1288743  48767.0                 108.0   
3                      764586      0.0                 106.0   
4                      536307  36357.0                  68.0   
...                       ...      ...                   ...   
25214             23405.98798  71941.0                  73.0   
25215             3096.511217  21604.0                  79.0   
25216             8163.890428      0.0                  85.0   
25217             7524.442436  21941.0                  96.0   
25218             2611.836866      0.0                  77.0   

      number_of_open_complaints     policy_type  vehicle_class  \
0                             0   Personal Auto  Four-Door Car   
1                             0   Personal Auto  Four-Door Car   
2                             0   Personal Auto   Two-Door Car   
3                             0  Corporate Auto            SUV   
4                             0   Personal Auto  Four-Door Car   
...                         ...             ...            ...   
25214                         0   Personal Auto  Four-Door Car   
25215                         0  Corporate Auto  Four-Door Car   
25216                         3  Corporate Auto  Four-Door Car   
25217                         0   Personal Auto  Four-Door Car   
25218                         0  Corporate Auto   Two-Door Car   

       total_claim_amount  
0                2.000000  
1             1131.000000  
2              566.000000  
3              529.000000  
4               17.000000  
...                   ...  
25214          198.234764  
25215          379.200000  
25216          790.784983  
25217          691.200000  
25218          369.600000  

[25219 rows x 11 columns]

In [37]:
df_final = functions.cleaning_invalid_values(df_final)

df_final = functions.formatting_data_types(df_final)

df_final = functions.dealing_with_null_values(df_final)

df_final = functions.dealing_with_duplicates(df_final)

df_final = df_final.reset_index(drop=True)

df_final.head()


customer       state gender             education  customer_lifetime_value  \
0  RB50392  Washington      U                Master              7714.876863   
1  QZ44356     Arizona      F              Bachelor              7714.876863   
2  AI49188      Nevada      F              Bachelor              7714.876863   
3  WW63253  California      M              Bachelor              7714.876863   
4  GA49547  Washington      M  High School or Below              7714.876863   

    income  monthly_premium_auto  number_of_open_complaints     policy_type  \
0      0.0                1000.0                        0.0   Personal Auto   
1      0.0                  94.0                        0.0   Personal Auto   
2  48767.0                 108.0                        0.0   Personal Auto   
3      0.0                 106.0                        0.0  Corporate Auto   
4  36357.0                  68.0                        0.0   Personal Auto   

   vehicle_class  total_claim_amount  
0  Four-Door Car                 2.0  
1  Four-Door Car              1131.0  
2   Two-Door Car               566.0  
3            SUV               529.0  
4  Four-Door Car                17.0

# Challenge 2: Structuring Data

In this challenge, we will continue to work with customer data from an insurance company, but we will use a dataset with more columns, called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by performing data cleaning, formatting, and structuring.

In [40]:
# Your code goes here
df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv")

df.head(10)

unnamed:_0 customer       state  customer_lifetime_value response  \
0           0  DK49336     Arizona              4809.216960       No   
1           1  KX64629  California              2228.525238       No   
2           2  LZ68649  Washington             14947.917300       No   
3           3  XL78013      Oregon             22332.439460      Yes   
4           4  QA50777      Oregon              9025.067525       No   
5           5  HO81102  California              4745.181764       No   
6           6  IW72280  California              5035.035257       No   
7           7  IH64929  California              4956.247235       No   
8           8  FM55990  California              5989.773931      Yes   
9           9  QX31376      Oregon             17114.328150       No   

   coverage             education effective_to_date employmentstatus gender  \
0     Basic               College        2011-02-18         Employed      M   
1     Basic               College        2011-01-18       Unemployed      F   
2     Basic              Bachelor        2011-02-10         Employed      M   
3  Extended               College        2011-01-11         Employed      M   
4   Premium              Bachelor        2011-01-17    Medical Leave      F   
5     Basic  High School or Below        2011-02-14         Employed      M   
6     Basic                Doctor        2011-02-14         Employed      F   
7     Basic  High School or Below        2011-02-24         Employed      M   
8   Premium               College        2011-01-19         Employed      M   
9  Extended                Master        2011-01-04         Employed      F   

   ...  number_of_policies     policy_type        policy  renew_offer_type  \
0  ...                   9  Corporate Auto  Corporate L3            Offer3   
1  ...                   1   Personal Auto   Personal L3            Offer4   
2  ...                   2   Personal Auto   Personal L3            Offer3   
3  ...                   2  Corporate Auto  Corporate L3            Offer2   
4  ...                   7   Personal Auto   Personal L2            Offer1   
5  ...                   7   Personal Auto   Personal L3            Offer1   
6  ...                   4  Corporate Auto  Corporate L2            Offer2   
7  ...                   3   Personal Auto   Personal L2            Offer2   
8  ...                   1   Personal Auto   Personal L1            Offer2   
9  ...                   2  Corporate Auto  Corporate L3            Offer2   

   sales_channel  total_claim_amount  vehicle_class  vehicle_size  \
0          Agent          292.800000  Four-Door Car       Medsize   
1    Call Center          744.924331  Four-Door Car       Medsize   
2    Call Center          480.000000            SUV       Medsize   
3         Branch          484.013411  Four-Door Car       Medsize   
4         Branch          707.925645  Four-Door Car       Medsize   
5          Agent          292.800000  Four-Door Car       Medsize   
6         Branch          287.556107  Four-Door Car       Medsize   
7         Branch          247.402353   Two-Door Car       Medsize   
8         Branch          739.200000     Sports Car       Medsize   
9          Agent          408.000000  Four-Door Car         Small   

  vehicle_type month  
0            A     2  
1            A     1  
2            A     2  
3            A     1  
4            A     1  
5            A     2  
6            A     2  
7            A     2  
8            A     1  
9            A     1  

[10 rows x 27 columns]

In [41]:
df = df.drop('unnamed:_0', axis=1)

After checking the column names I decided to drop the 'unnamed:_0' column which seemed to be a duplicate of the label column. The column names appear to be correctly formated.

In [42]:
display(df.dtypes)

customer                          object
state                             object
customer_lifetime_value          float64
response                          object
coverage                          object
education                         object
effective_to_date                 object
employmentstatus                  object
gender                            object
income                             int64
location_code                     object
marital_status                    object
monthly_premium_auto               int64
months_since_last_claim          float64
months_since_policy_inception      int64
number_of_open_complaints        float64
number_of_policies                 int64
policy_type                       object
policy                            object
renew_offer_type                  object
sales_channel                     object
total_claim_amount               float64
vehicle_class                     object
vehicle_size                      object
vehicle_type    

In [43]:
df["months_since_last_claim"]=df["months_since_last_claim"].astype('int64')
display(df.dtypes)

customer                          object
state                             object
customer_lifetime_value          float64
response                          object
coverage                          object
education                         object
effective_to_date                 object
employmentstatus                  object
gender                            object
income                             int64
location_code                     object
marital_status                    object
monthly_premium_auto               int64
months_since_last_claim            int64
months_since_policy_inception      int64
number_of_open_complaints        float64
number_of_policies                 int64
policy_type                       object
policy                            object
renew_offer_type                  object
sales_channel                     object
total_claim_amount               float64
vehicle_class                     object
vehicle_size                      object
vehicle_type    

In [44]:
display(df["number_of_open_complaints"].unique())
display(df["number_of_open_complaints"].value_counts())

array([0.        , 0.38425611, 3.        , 1.        , 2.        ,
       4.        , 5.        ])

0.000000    8160
1.000000    1145
0.384256     633
2.000000     414
3.000000     324
4.000000     166
5.000000      68
Name: number_of_open_complaints, dtype: int64

In [45]:
df["number_of_open_complaints"]=df["number_of_open_complaints"].astype('int64')
display(df["number_of_open_complaints"].value_counts())
df["number_of_open_complaints"].unique()

0    8793
1    1145
2     414
3     324
4     166
5      68
Name: number_of_open_complaints, dtype: int64

array([0, 3, 1, 2, 4, 5])

Checking the data types I decided to change the column "months_since_last_claim" and "number_of_open_complaints" to integer since they are numerical discrete data types. The rest of the data types seem to be correct for the column names.

In [46]:
display(df.isna().sum())

customer                         0
state                            0
customer_lifetime_value          0
response                         0
coverage                         0
education                        0
effective_to_date                0
employmentstatus                 0
gender                           0
income                           0
location_code                    0
marital_status                   0
monthly_premium_auto             0
months_since_last_claim          0
months_since_policy_inception    0
number_of_open_complaints        0
number_of_policies               0
policy_type                      0
policy                           0
renew_offer_type                 0
sales_channel                    0
total_claim_amount               0
vehicle_class                    0
vehicle_size                     0
vehicle_type                     0
month                            0
dtype: int64

There appears to be no missing values.

In [47]:
display(df.duplicated().sum())
df.drop_duplicates(keep='first', inplace=True)
display(df.duplicated().sum())

460

0

In [48]:
display(df.duplicated(subset="customer").sum())

1316

There is a large amount of values from the "customer" column that are the same, and seeing that this column should probably contain a unique id for the customers and that that amount can have a significant impact in the data that we extract from the dataframe, I decided to combine the first three columns of the dataframe to filter out possible duplicates.

In [51]:
display(df["customer"].value_counts())
df.drop_duplicates(subset=["customer","state","customer_lifetime_value"], keep='first', inplace=True)
display(df.duplicated(subset="customer").sum())

IA63417    2
BX70848    2
UK86050    2
IV32877    2
UY85263    2
          ..
YP19985    1
LM97847    1
ZI51939    1
PV77906    1
WA60547    1
Name: customer, Length: 9134, dtype: int64

370

1. You work at the marketing department and you want to know which sales channel brought the most sales in terms of total revenue. Using pivot, create a summary table showing the total revenue for each sales channel (branch, call center, web, and mail).
Round the total revenue to 2 decimal points.  Analyze the resulting table to draw insights.

In [52]:
groupby = df.groupby(["sales_channel"]).agg({"customer_lifetime_value": sum}).round(2).reset_index()
groupby

sales_channel  customer_lifetime_value
0         Agent              28685435.93
1        Branch              21611380.10
2   Call Center              14848863.09
3           Web              10863006.62

From analyzing the data, the highest revenue came from the Agent sales channel with a total claim amount of around 28685435.93 and the lowest came from the Web sales channel with 10863006.62.

2. Create a pivot table that shows the average customer lifetime value per gender and education level. Analyze the resulting table to draw insights.

In [53]:
df.groupby(["gender","education"]).agg({"customer_lifetime_value": "mean"}).round(2).reset_index()

gender             education  customer_lifetime_value
0      F              Bachelor                  7974.11
1      F               College                  7740.49
2      F                Doctor                  7292.36
3      F  High School or Below                  8585.56
4      F                Master                  8286.02
5      M              Bachelor                  7711.20
6      M               College                  7972.80
7      M                Doctor                  7529.84
8      M  High School or Below                  8024.68
9      M                Master                  8168.38

By analyzing the table we can see the highest average customer lifetime value corresponds to females with high school or below level of education with an average of around 8585.56 and the lowest average corresponds to females with doctorate degree with an average of around 7529.84.

## Bonus

You work at the customer service department and you want to know which months had the highest number of complaints by policy type category. Create a summary table showing the number of complaints by policy type and month.
Show it in a long format table.

*In data analysis, a long format table is a way of structuring data in which each observation or measurement is stored in a separate row of the table. The key characteristic of a long format table is that each column represents a single variable, and each row represents a single observation of that variable.*

*More information about long and wide format tables here: https://www.statology.org/long-vs-wide-data/*

In [ ]:
# Your code goes here